### PR TITLE
Add local-first MLX search workflow

### PR DESCRIPTION
**Pull request by sueun-dev**
## Summary
- add runtime shard and validation-sequence limits for local proxy iteration without changing the default challenge path
- add a local MLX sweep harness with proxy/full-val modes and budget-aware architecture filtering
- document the Apple Silicon local search workflow in the README and a dedicated playbook

## Why
Local Apple Silicon iteration was bottlenecked by full validation on every run and by fully manual hyperparameter sweeps. This patch keeps the default challenge behavior intact while making local search substantially faster and easier to compare.

## Validation
- python3 -m py_compile train_gpt.py train_gpt_mlx.py tools/local_search.py
- exercised tools/local_search.py on Apple Silicon with proxy-search sessions

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Add local-first MLX search workflow

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


## Analysis

**Head SHA:** ef03d329bc7646b63b1afd7c48ec29bbc47000e1

**Files changed:** README.md, docs/local_search.md, tools/local_search.py, train_gpt.py, train_gpt_mlx.py

### Eval path (train_gpt.py lines 231–290)

The `eval_val` function is the sole scoring path. It is fully wrapped in `torch.inference_mode()` (line 262) and iterates over `val_tokens` in fixed batches with no gradient computation, no optimizer steps, and no weight mutation:

```python
model.eval()
with torch.inference_mode():
    for batch_seq_start in range(seq_start, seq_end, local_batch_seqs):
        ...
        batch_loss = model(x, y).detach()
```

`eval_val` is called twice: once per validation interval during training (line 1014) and once post-quantization for the roundtrip check (line 1136). Neither call modifies model weights.

### N-gram family bug check
No hash tables, no n-gram caches, no prime-multiplied key construction anywhere in the file. The only use of "primes" is in a comment at line 963 about warming up optimizer paths. CLEAN.

### TTT / Pre-Quant TTT check
No `ttt_adapt`, no per-chunk adaptation loop, no multi-epoch AdamW over `val_tokens`. The only optimizers present are the training-phase optimizers (Adam + Muon), which are not called during or after `eval_val`. CLEAN.

### SLOT check
No per-window NLL mask optimization pattern. No scored-region AdamW calls on validation positions. CLEAN.

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
